pallas-txbuilder/src/transaction/model.rs

```diff
@@ -0,0 +1,712 @@
+                    })
+                    .or_insert(amount);
+            })
+            .or_insert_with(|| {
+                let mut map: HashMap<Bytes, u64> = HashMap::new();
+                map.insert(name.clone().into(), amount);
+                map
+            });
+
+        self.assets = Some(OutputAssets(assets));
+
+        Ok(self)
+    }
+
+    pub fn set_inline_datum(mut self, plutus_data: Vec<u8>) -> Self {
+        self.datum = Some(Datum {
+            kind: DatumKind::Inline,
+            bytes: plutus_data.into(),
+        });
+
+        self
+    }
+
+    pub fn set_datum_hash(mut self, datum_hash: Hash<32>) -> Self {
+        self.datum = Some(Datum {
+            kind: DatumKind::Inline,
+            bytes: datum_hash.to_vec().into(),
+        });
+
+        self
+    }
+
+    pub fn set_inline_script(mut self, language: ScriptKind, bytes: Vec<u8>) -> Self {
+        self.script = Some(Script {
+            kind: language,
+            bytes: bytes.into(),
+        });
+
+        self
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct OutputAssets(HashMap<PolicyId, HashMap<AssetName, u64>>);
+
+impl Deref for OutputAssets {
+    type Target = HashMap<PolicyId, HashMap<Bytes, u64>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl OutputAssets {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    pub fn from_map(map: HashMap<PolicyId, HashMap<Bytes, u64>>) -> Self {
+        Self(map)
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Default)]
+pub struct MintAssets(HashMap<PolicyId, HashMap<AssetName, i64>>);
+
+impl Deref for MintAssets {
+    type Target = HashMap<PolicyId, HashMap<Bytes, i64>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl MintAssets {
+    pub fn new() -> Self {
+        MintAssets(HashMap::new())
+    }
+
+    pub fn from_map(map: HashMap<PolicyId, HashMap<Bytes, i64>>) -> Self {
+        Self(map)
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone, Copy)]
+#[serde(rename_all = "snake_case")]
+pub enum ScriptKind {
+    Native,
+    PlutusV1,
+    PlutusV2,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct Script {
+    pub kind: ScriptKind,
+    pub bytes: ScriptBytes,
+}
+
+impl Script {
+    pub fn new(kind: ScriptKind, bytes: Vec<u8>) -> Self {
+        Self {
+            kind,
+            bytes: bytes.into(),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum DatumKind {
+    Hash,
+    Inline,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct Datum {
+    pub kind: DatumKind,
+    pub bytes: DatumBytes,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+pub enum RedeemerPurpose {
+    Spend(Input),
+    Mint(PolicyId),
+    // Reward TODO
+    // Cert TODO
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct ExUnits {
```

Comment:
potential candidate for `pallas-plutus`? We have the same type in the Aiken uplc crate so we can use the same one in both situations